hr_planning_resources/models/hr_leave.py

```diff
@@ -0,0 +1,188 @@
+from collections import defaultdict
+from datetime import timedelta
+from functools import lru_cache
+from itertools import groupby
+
+from pytz import timezone, utc
+
+from odoo import _, api, models
+from odoo.tools.misc import get_lang
+
+
+def format_time(env, time):
+    return time.strftime(get_lang(env).time_format)
+
+
+def format_date(env, date):
+    return date.strftime(get_lang(env).date_format)
+
+
+class HrLeave(models.Model):
+    _inherit = "hr.leave"
+
+    @api.model
+    def _get_leaves(self, date_from, date_to, employee_ids):
+        calendar_leaves = self._get_calendar_leaves(date_from, date_to, employee_ids)
+        leaves = self._group_leaves_by_employee(calendar_leaves, employee_ids)
+        hr_leaves = self._get_hr_leaves(date_from, date_to, employee_ids)
+        for leave in hr_leaves:
+            leaves[leave.employee_id.id].append(leave)
+        return leaves
+
+    def _get_calendar_leaves(self, date_from, date_to, employee_ids):
+        return self.env["resource.calendar.leaves"].search(
+            [
+                ("time_type", "=", "leave"),
+                "|",
+                ("company_id", "in", employee_ids.mapped("company_id").ids),
+                ("company_id", "=", False),
+                "|",
+                ("resource_id", "in", employee_ids.mapped("resource_id").ids),
+                ("resource_id", "=", False),
+                ("date_from", "<=", date_to),
+                ("date_to", ">=", date_from),
+            ],
+            order="date_from",
+        )
+
+    def _group_leaves_by_employee(self, calendar_leaves, employee_ids):
+        leaves = defaultdict(list)
+        for leave in calendar_leaves:
+            for employee in employee_ids:
+                if self._is_leave_relevant_to_employee(leave, employee):
+                    leaves[employee.id].append(leave)
+        return leaves
+
+    def _is_leave_relevant_to_employee(self, leave, employee):
+        return (
+            (not leave.company_id or leave.company_id == employee.company_id)
+            and (not leave.resource_id or leave.resource_id == employee.resource_id)
+            and (
+                not leave.calendar_id
+                or leave.calendar_id == employee.resource_calendar_id
+            )
+        )
+
+    def _get_hr_leaves(self, date_from, date_to, employee_ids):
+        return self.env["hr.leave"].search(
+            [
+                ("employee_id", "in", employee_ids.ids),
+                ("state", "in", ["confirm", "validate1"]),
+                ("date_from", "<=", date_to),
+                ("date_to", ">=", date_from),
+            ],
+            order="date_from",
+        )
+
+    def _get_leave_message_warning(self, leaves, employee, date_from, date_to):
+        @lru_cache(None)
+        def localize(date):
+            return (
+                utc.localize(date)
+                .astimezone(timezone(self.env.user.tz or "UTC"))
```

Comment:
You already have `fields.Datetime.context_timestamp(...)` who is doing a part of your code